### PR TITLE
Meshlc missing pointmatch tolerant

### PR DESCRIPTION
**Pull request by djkapner**
In the event that a few tilepairs don't have matches due to bloodvessels, this prevents a failure. Since there are >300 tilepairs per reference set, there's generally quite a bit of data left.

### CODE REVIEW

**Comment by djkapner** [comment]
I think a way to add this to a test is to add a couple of dummy/empty images to the test data, and add those to the tilepair file. I could do that, it just wasn't clear what that test would prove: that empty images don't give pointmatches?

**Comment by fcollman** [comment]
Just that your old code failed in this condition and your new code doesn’t. If we refactored the innards at some point of how this code works it captures the use case that we forgot the first time it was written.

File: rendermodules/mesh_lens_correction/LensPointMatches.py [hunk 1]
```diff
@@ -184,24 +106,29 @@ def find_matches(fargs):
...
+
+        #render = renderapi.connect(**args['render'])
+        render = args['render']
+        pm_dict = make_pm(ids, gids, k1, k2)
+        renderapi.pointmatch.import_matches(
+          args['match_collection'],
+          [pm_dict],
+          render=render)
+    else:
+        print('no pointmatches found for tilepair:')
```
Comment (fcollman) [inline rendermodules/mesh_lens_correction/LensPointMatches.py]: this should be a logging statement and not a print statement

File: rendermodules/mesh_lens_correction/MeshAndSolveTransform.py [hunk 2]
```diff
@@ -482,6 +461,23 @@ def solve(self):
...
+                  (v, e.min(), e.max(), e.mean(), e.std()))
+
+        scale = np.array([tf.scale for tf in self.transforms])
+        message += ('xscale: [%8.2f,%8.2f] %8.2f+/-%8.2f\n' %
+                    (scale[:, 0].min(), scale[:, 0].max(),
+                     scale[:, 0].mean(), scale[:, 0].std()))
+        message += ('yscale: [%8.2f,%8.2f] %8.2f+/-%8.2f\n' %
+                    (scale[:, 1].min(), scale[:, 1].max(),
+                     scale[:, 1].mean(), scale[:, 1].std()))
+        print(message)
```
Comment (fcollman) [inline rendermodules/mesh_lens_correction/MeshAndSolveTransform.py]: also should be a logging statement

File: rendermodules/mesh_lens_correction/MeshAndSolveTransform.py [hunk 3]
```diff
@@ -620,7 +616,6 @@ def MeshAndSolve(self, render,
         print('solving...')
         self.solve()
 
-        #self.report_solution()
         print('calling ThinPlateClient to create dataString')
```
Comment (fcollman) [inline rendermodules/mesh_lens_correction/MeshAndSolveTransform.py]: ditto about logging.. probably remove this actually

File: rendermodules/mesh_lens_correction/LensPointMatches.py [hunk 4]
```diff
@@ -263,22 +190,22 @@ def match_image_pairs(render, tilespecs, match_collection, downsample_scale, nfe
...
 
-    if index_list is None:
-        index_list = range(tile_index.shape[0])
+        if index_list is None:
+            index_list = range(tile_index.shape[0])
     
-    for i in index_list:
-        impaths = [t.ip.mipMapLevels[0].imageUrl.split('file://')[-1]
+        for i in index_list:
+            impaths = [t.ip.mipMapLevels[0].imageUrl.split('file://')[-1]
```
Comment (RussTorres) [inline rendermodules/mesh_lens_correction/LensPointMatches.py]: This is fine for a raw data stack, but the current ImagePyramid interface does not guarantee an ordering of mipMapLevels (which are in a list).  In addition, stripping the `file://` lead can cause failures. a more stable way to get the path to the file (for now) is `urllib.unquote(urlparse.urlparse(t.ip.mipMapLevels.get(0)['imageUrl']).path)`.

File: integration_tests/test_mesh_lens_correction.py [hunk 5]
```diff
@@ -1,72 +1,63 @@
...
-import logging
 import renderapi
 import os
-import numpy as np
 
 from rendermodules.mesh_lens_correction.do_mesh_lens_correction import MeshLensCorrection
 from rendermodules.mesh_lens_correction.LensPointMatches import generate_point_matches
 from test_data import render_params, TEST_DATA_ROOT
 
 example = {
```
[the remark this answers is not in the document]
Comment (mahaling) [reply integration_tests/test_mesh_lens_correction.py]: I disagree... Its better to get the example json input in the tests own files as the test_data.py is getting longer and longer day by day. The templated json files are anyway in the test_files folder. 

Any changes to the environment variable in test_data.py will reflect here as well without any changes need to be done to this test script.

What needs to be included is comprehensive test cases for lens correction that resolves the issues recorded in this repo.

**Comment by djkapner** [comment]
so, if I want to guarantee to retrieve the mipmap level 0 from an ip, do i need to parse the paths? there's nothing else in that data structure, right?
```

mipmapLevels |  
-- | --
0 |  
imageUrl | "file:///allen/programs/celltypes/production/wijem/long_term/em_montage_247488_8R_z0_2018_02_14_00_21_30_00_00/20180213163908673_247488_8R_tape070A_01_000000_0_61_34.tif"
1 |  
imageUrl | "file:/allen/programs/celltypes/production/wijem/mipmap/em_montage_247488_8R_z0_2018_02_14_00_21_30_00_00/1/allen/programs/celltypes/production/incoming/wijem/247488_8R_tape070A_01_000000_0/20180213163908673_247488_8R_tape070A_01_000000_0_61_34.tif.tif"
2 |  
imageUrl | "file:/allen/programs/celltypes/production/wijem/mipmap/em_montage_247488_8R_z0_2018_02_14_00_21_30_00_00/2/allen/programs/celltypes/production/incoming/wijem/247488_8R_tape070A_01_000000_0/20180213163908673_247488_8R_tape070A_01_000000_0_61_34.tif.tif"
3 |  
imageUrl | "file:/allen/programs/celltypes/production/wijem/mipmap/em_montage_247488_8R_z0_2018_02_14_00_21_30_00_00/3/allen/programs/celltypes/production/incoming/wijem/247488_8R_tape070A_01_000000_0/20180213163908673_247488_8R_tape070A_01_000000_0_61_34.tif.tif"
4 |  
imageUrl | "file:/allen/programs/celltypes/production/wijem/mipmap/em_montage_247488_8R_z0_2018_02_14_00_21_30_00_00/4/allen/programs/celltypes/production/incoming/wijem/247488_8R_tape070A_01_000000_0/20180213163908673_247488_8R_tape070A_01_000000_0_61_34.tif.tif"
5 |  
imageUrl | "file:/allen/programs/celltypes/production/wijem/mipmap/em_montage_247488_8R_z0_2018_02_14_00_21_30_00_00/5/allen/programs/celltypes/production/incoming/wijem/247488_8R_tape070A_01_000000_0/20180213163908673_247488_8R_tape070A_01_000000_0_61_34.tif.tif"
6 |  
imageUrl | "file:/allen/programs/celltypes/production/wijem/mipmap/em_montage_247488_8R_z0_2018_02_14_00_21_30_00_00/6/allen/programs/celltypes/production/incoming/wijem/247488_8R_tape070A_01_000000_0/20180213163908673_247488_8R_tape070A_01_000000_0_61_34.tif.tif"
```

File: integration_tests/test_mesh_lens_correction.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (djkapner) [reply integration_tests/test_mesh_lens_correction.py]: i think issues like logging and improved integration tests should be another branch/PR. This one is just to unclog those 5 hung corrections.

File: rendermodules/mesh_lens_correction/LensPointMatches.py (hunk 4 above)
Comment (djkapner) [reply rendermodules/mesh_lens_correction/LensPointMatches.py]: so, if I want to guarantee to retrieve the mipmap level 0 from an ip, do i need to parse the paths? there's nothing else in that data structure, right?
```

mipmapLevels |  
-- | --
0 |  
imageUrl | "file:///allen/programs/celltypes/production/wijem/long_term/em_montage_247488_8R_z0_2018_02_14_00_21_30_00_00/20180213163908673_247488_8R_tape070A_01_000000_0_61_34.tif"
1 |  
imageUrl | "file:/allen/programs/celltypes/production/wijem/mipmap/em_montage_247488_8R_z0_2018_02_14_00_21_30_00_00/1/allen/programs/celltypes/production/incoming/wijem/247488_8R_tape070A_01_000000_0/20180213163908673_247488_8R_tape070A_01_000000_0_61_34.tif.tif"
2 |  
imageUrl | "file:/allen/programs/celltypes/production/wijem/mipmap/em_montage_247488_8R_z0_2018_02_14_00_21_30_00_00/2/allen/programs/celltypes/production/incoming/wijem/247488_8R_tape070A_01_000000_0/20180213163908673_247488_8R_tape070A_01_000000_0_61_34.tif.tif"
3 |  
imageUrl | "file:/allen/programs/celltypes/production/wijem/mipmap/em_montage_247488_8R_z0_2018_02_14_00_21_30_00_00/3/allen/programs/celltypes/production/incoming/wijem/247488_8R_tape070A_01_000000_0/20180213163908673_247488_8R_tape070A_01_000000_0_61_34.tif.tif"
4 |  
imageUrl | "file:/allen/programs/celltypes/production/wijem/mipmap/em_montage_247488_8R_z0_2018_02_14_00_21_30_00_00/4/allen/programs/celltypes/production/incoming/wijem/247488_8R_tape070A_01_000000_0/20180213163908673_247488_8R_tape070A_01_000000_0_61_34.tif.tif"
5 |  
imageUrl | "file:/allen/programs/celltypes/production/wijem/mipmap/em_montage_247488_8R_z0_2018_02_14_00_21_30_00_00/5/allen/programs/celltypes/production/incoming/wijem/247488_8R_tape070A_01_000000_0/20180213163908673_247488_8R_tape070A_01_000000_0_61_34.tif.tif"
6 |  
imageUrl | "file:/allen/programs/celltypes/production/wijem/mipmap/em_montage_247488_8R_z0_2018_02_14_00_21_30_00_00/6/allen/programs/celltypes/production/incoming/wijem/247488_8R_tape070A_01_000000_0/20180213163908673_247488_8R_tape070A_01_000000_0_61_34.tif.tif"
```

**Comment by RussTorres** [comment]
@djkapner I'm not sure what you mean by parse the paths in this context -- the url decoding libraries help by, e.g. changing `%20` to a space for the filesystem.  They also handle less common uris such as `file:/path/to/file`.

File: rendermodules/mesh_lens_correction/LensPointMatches.py (hunk 4 above)
Comment (RussTorres) [reply rendermodules/mesh_lens_correction/LensPointMatches.py]: I'm not sure what you mean by parse the paths in this context -- the url decoding libraries help by, e.g. changing %20 to a space for the filesystem. They also handle less common uris such as file:/path/to/file.

Comment (RussTorres) [reply rendermodules/mesh_lens_correction/LensPointMatches.py]: Ah, I had a typo in my command which has been edited.  the imagepyramid get method will search mipMapLevels by level.

Comment (djkapner) [reply rendermodules/mesh_lens_correction/LensPointMatches.py]: i was referring to `... the current ImagePyramid interface does not guarantee an ordering of mipMapLevels (which are in a list)...`
where I think you were referring to `t.ip.mipMapLevels[0]`
so, my question is, if I can't rely on the first element of the list being mipmap level 0, what's a good strategy?

Comment (djkapner) [reply rendermodules/mesh_lens_correction/LensPointMatches.py]: got it, ok.

Comment (mahaling) [reply rendermodules/mesh_lens_correction/LensPointMatches.py]: replacing that line with 
impaths = [urllib.unquote(urlparse.urlparse(str(t.ip.mipMapLevels[0].imageUrl)).path)
                    for t in tilespecs[tile_index[i]]]
should work with proper import statements for urllib and urlparse.
